javaslang/src/main/java/javaslang/collection/LeafType.java

```diff
@@ -5,99 +5,238 @@
+        final LeafType<T> type = of((Class<T>) primitiveClass);
+        final Object results = type.newInstance(array.length);
+        for (int i = 0; i < array.length; i++) {
+            type.setAt(results, i, array[i]);
+        }
+        return (T) results;
+    }
+
+    /* convert to primitive */
+    private static final Class<?>[] WRAPPERS = { Boolean.class, Byte.class, Character.class, Double.class, Float.class, Integer.class, Long.class, Short.class, Void.class };
+    private static final Class<?>[] PRIMITIVES = { boolean.class, byte.class, char.class, double.class, float.class, int.class, long.class, short.class, void.class };
+
+    static Class<?> asPrimitive(Class<?> wrapper) {
+        final int i = primitiveIndex(wrapper);
+        return (i < 0) ? wrapper
+                       : PRIMITIVES[i];
+    }
+
+    private static int primitiveIndex(Class<?> wrapper) { /* linear search is faster than binary search here */
+        for (int j = 0; j < WRAPPERS.length; j++) {
+            if (wrapper == WRAPPERS[j]) {
+                return j;
+            }
+        }
+        return -1;
+    }
+}
+
+final class ObjectArray extends LeafType<Object> {
+    private static final long serialVersionUID = 1L;
+    static final ObjectArray OBJECT_TYPE = new ObjectArray();
+
+    @Override
+    boolean isPrimitive() { return false; }
+
+    @Override
+    Class<Object> type() { return Object.class; }
+
+    static final Object[] EMPTY = new Object[0];
+
+    @Override
+    Object[] empty() { return EMPTY; }
+
+    private static Object[] cast(Object array) { return (array == null) ? EMPTY : (Object[]) array; }
+
+    @Override
+    int lengthOf(Object array) { return cast(array).length; }
+
+    @Override
+    Object getAt(Object array, int index) { return cast(array)[index]; }
+
+    @Override
+    void setAt(Object array, int index, Object value) { cast(array)[index] = value; }
+
+    @Override
+    Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size) {
         if (size == 0) {
             return new Object[arraySize];
         } else {
-            final Object result = new Object[arraySize];
-            System.arraycopy(source, sourceFrom, result, destinationFrom, size);
+            final Object[] result = new Object[arraySize];
+            System.arraycopy(array, sourceFrom, result, destinationFrom, size); // has to be near the object allocation to avoid zeroing out the array
             return result;
         }
     }
+}
 
-    /** array access, avoiding call-site casts */
-    @SuppressWarnings("unchecked")
-    static <T> T getAt(Object array, int index) { return (T) ((Object[]) array)[index]; }
-    @SuppressWarnings("unchecked")
-    static void setAt(Object array, int index, Object value) { ((Object[]) array)[index] = value; }
-    static int lengthOf(Object array) {
-        return ((Object[]) array).length;
+final class IntArray extends LeafType<Integer> {
```

Comment:
`int`s are used very often, let's optimize their access (storage is the same in all cases)